### DISCUSSION TITLE
Unrolling a sheet into a linear sequence

### DISCUSSION THREAD

**Discussion by infojunkie** [w3c-cg/musicxml#453] [Q&A] [answered]
I am wondering if there are published, well-known algorithms for "unrolling" a sheet music. By this I mean converting it to a linear sequence of musical events, without repetitions or loops - e.g. a MIDI stream.

I had to implement such an algorithm in my [MusicXML to MIDI accompaniment generator project](https://github.com/infojunkie/musicxml-mma). To help validate it, I added a fairly involved [test sheet](https://github.com/infojunkie/musicxml-mma/blob/main/test/data/repeats.musicxml) and tested my output against MuseScore's. But having a published reference would be even better in terms of validation and comprehensiveness.

**Comment by mdgood** [comment]
I don't know of any such publication offhand. Since MuseScore code is open source, it seems it could be possible to write up their algorithm somewhere, with appropriate credit.

**Reply by benspratling4** [reply]
@infojunkie  Did you end up finding one?  I've spent the last 6 months banging my head against the wall trying to figure out how to figure out who sings what when, and right now my algorithm needs 9 passes over the song (and 15 before I'm done), and needs to do statistical analyses and I still need to build a wavelet analysis over lyric name / numbers to finish it off.  My product will live or die on my getting this perfectly correct without a musician in the loop, to minimize the cost to publishers.

**Reply by infojunkie** [reply]
Yes, I [made an XSL script](https://github.com/infojunkie/musicxml-mma) that unrolls the MusicXML repeats and jumps into another format that I use for MIDI generation. The relevant logic is in the [template that processes  measures](https://github.com/infojunkie/musicxml-mma/blob/main/musicxml-mma.xsl#L93). It currently needs a single pass with lots of look-ahead and look-behind, and is probably very buggy :smile:

**Reply by infojunkie** [reply, marked as answer]
[Here's an XSL script](https://github.com/infojunkie/musicxml-mma/blob/87119999d3953b55fbcc2f6224c68eb5a26708bc/musicxml-unroll.xsl) that unrolls a MusicXML file to another MusicXML file that contains no jumps or repeats. I would love if someone else could test it.
